Fix DataPackUtils::processMCMeta error

dsch7705 opened the pull request:
Fixes #4331 

Occasionally, a resource/ data pack will have malformed JSON data in its pack.mcmeta in the form of credits or other information that the JSON parser collides with. This can be fixed by trimming the raw_data that is passed to processMCMeta before parsing such that only the top level JSON object remains.

dsch7705 committed:
Trim raw_data passed to processMCMeta

Signed-off-by: Dylan Schooner <dschooner05@gmail.com>

getchoo reviewed (commented):
I'm not really sure I like this idea honestly

People adding credits like this to make their pack.mcmeta file is...bad practice to say the least. I'm frankly surprised it's able to be uploaded to Modrinth or works correctly with the game and other launchers (does it...?)

Silently allowing this kind of thing also feels very wrong and like one of those "feature landmines" so to speak - that become a big pain in any refactorings or just trying to make sense of code. It also becomes a bit of a slippery slope: sure it works on this occasion, but what if someone decides to place the credits at the top of the file and it happens to use a brace? What if they add comments inside of the JSON data itself?

IMO the correct solution to these types of problems is to inform mod authors that what they're doing probably is not the best idea, and encourage them to include credits in another (safer) way. Otherwise, this may catch on and strange fixes like this will need to be copied evermore in similar applications

The code itself LGTM, though. Just my ramble :)

dsch7705 commented:
> I'm not really sure I like this idea honestly
> 
> People adding credits like this to make their pack.mcmeta file is...bad practice to say the least. I'm frankly surprised it's able to be uploaded to Modrinth or works correctly with the game and other launchers (does it...?)
> 
> Silently allowing this kind of thing also feels very wrong and like one of those "feature landmines" so to speak - that become a big pain in any refactorings or just trying to make sense of code. It also becomes a bit of a slippery slope: sure it works on this occasion, but what if someone decides to place the credits at the top of the file and it happens to use a brace? What if they add comments inside of the JSON data itself?
> 
> IMO the correct solution to these types of problems is to inform mod authors that what they're doing probably is not the best idea, and encourage them to include credits in another (safer) way. Otherwise, this may catch on and strange fixes like this will need to be copied evermore in similar applications
> 
> The code itself LGTM, though. Just my ramble :)

I absolutely agree, this "fixes" one thing but like you said, there are other ways for this to break. It honestly surprises me that Minecraft itself accepts this type of format without issue, IMO they should adhere to the JSON standard.

Trial97 commented on launcher/minecraft/mod/tasks/LocalDataPackParseTask.cpp:
```diff
@@ -176,6 +176,23 @@ bool processZIP(DataPack* pack, ProcessingLevel level)
+            open_braces++;
+        else if (byte == '}')
+            close_braces++;
+    }
```
try:
```suggestion
    int braces = 0;
    qsizetype valid_bytes = 0;
    for (auto& byte : raw_data) {
        if (valid_bytes > 0 && braces == 0)
            break;

        valid_bytes++;
        if (byte == '{')
            braces++;
        else if (byte == '}')
            braces--;
    }
```

dsch7705 replied:
This is cleaner, thank you. I feel conflicted making further commits though, seeing as allowing such formatting (including credits, etc) is fragile and probably not of best practice. Thoughts?

Trial97 replied:
In my opinion, we would still need to do our best to decode it.
Even if the JSON is not valid, we should at least display the file as incompatible, not hide it completely.
If the game sees this as a valid data pack, I see no reason for Prism not to at least try to match it.

rosembackpedrob replied:
This seem like the best compromise. Informing something like "Resource Pack Parsing Error" would atleast communicate that it may be a fault of bad practices on the end of the resourcepack.

rosembackpedrob commented:
Good work on the quick fix. And I do agree with what you said about it being best to report for the resource pack author. I did added an issue, even though I am not sure if the project is still active.
As i said in the issue, I already believed it was some parsing preventing UI to update properly, just thoght it was on Prism's end. Considering it's not it may be optimal to not merge this. And in case of any other similar issues submitted in the future, people could just redirect to this as info. But ultimately, I think that it would be better if Minecraft or Modrinth didn't accepted it. As it would be much clearer to end user (and developers) that this is not encouraged approach to JSONs. It's suboptimal if loaders like this need to do all this work just to fix something that is not standard at all in JSON usage.

dsch7705 committed:
Refactor brace counting

Signed-off-by: Dylan Schooner <dschooner05@gmail.com>

TheKodeToad commented:
> It also becomes a bit of a slippery slope: sure it works on this occasion, but what if someone decides to place the credits at the top of the file and it happens to use a brace? What if they add comments inside of the JSON data itself?

Minecraft (at least nowadays) uses strict JSON parsing (Strictness.STRICT). I think it just doesn't check whether there's still some stuff in the stream after parsing the JSON, and the JSON library probably doesn't check that for flexibility; Minecraft should be checking whether there are remaining bytes in the file and throw an error if that's the case.

If I add a comment, I get
```
[11:33:43] [Render thread/ERROR]: Couldn't load file/Re-covered pack metadata: com.google.gson.stream.MalformedJsonException: Use JsonReader.setStrictness(Strictness.LENIENT) to accept malformed JSON at line 2 column 2 path $.
```
and the pack doesn't appear

For older versions, this potentially would be a slippery slope as I think older versions of GSON didn't allow you to fully disable quirks.

Trial97 reviewed (approved):
This is still fragile, but at least now, this should be on par with the game.
Again, the better option would still be to report to the user the failure somehow.

TheKodeToad reviewed (changes requested):
This implementation is too fragile for my liking... braces inside strings could confuse it.
I don't think there's really a good solution for this, unforunately. I mean you *can* recognise strings but it won't be pretty.

dsch7705 commented:
> This is still fragile, but at least now, this should be on par with the game. Again, the better option would still be to report to the user the failure somehow.

I'm going to revert the previous commit, its an extra variable but the tests are failing I assume because the first character in the test data is a newline, which should still parse just fine. Only keeping track of the balance fails when the first character is not '{'.

dsch7705 commented:
> This implementation is too fragile for my liking... braces inside strings could confuse it. I don't think there's really a good solution for this, unforunately. I mean you _can_ recognise strings but it won't be pretty.

This is true.

TheKodeToad commented:
> This is still fragile, but at least now, this should be on par with the game.

That's the problem, it's not

Say you add a smiley face to your pack
```json
{
    "pack": {
        "description": "My cool pack :}"
    }
}
```
It would be pretty confusing that it disappears from prism (and yep it does appear in minecraft)

dsch7705 committed:
Revert "Refactor brace counting"

This reverts commit 69145779e3ee51db165fb9cc9d373a3e44fac052.

dsch7705 commented:
In my opinion it may be better to simply exclude the meta data and inform the user of a misformed pack.mcmeta file. That way at least the pack still shows up on the list.

TheKodeToad commented:
Yes, I think that'd be good! I can't remember if there was a specific reason to do this for resource packs? But IIRC it originated from trying to hide shader configs.

dsch7705 commented:
> Yes, I think that'd be good! I can't remember if there was a specific reason to do this for resource packs? But IIRC it originated from trying to hide shader configs.

See but the pack.mcmeta is the only thing that is used to verify that a zip/ folder is a valid pack.

dsch7705 committed:
Display packs regardless of pack.mcmeta validity

Remove data trimming

Signed-off-by: Dylan Schooner <dschooner05@gmail.com>

dsch7705 committed:
DCO Remediation Commit for Dylan Schooner <dschooner05@gmail.com>

I, Dylan Schooner <dschooner05@gmail.com>, hereby add my Signed-off-by to this commit: 2e9e0550a2327f67287ed81242c42ac1a7b40682

Signed-off-by: Dylan Schooner <dschooner05@gmail.com>

dsch7705 commented:
<img width="1101" height="743" alt="bad_mcmeta2" src="https://github.com/user-attachments/assets/73c1ac71-bb03-4a77-9a3d-6f9a8dc39a79" />

This is what I have come up with so far.

TheKodeToad commented on launcher/minecraft/mod/ResourcePackFolderModel.cpp:
```diff
@@ -102,6 +104,9 @@ QVariant ResourcePackFolderModel::data(const QModelIndex& index, int role) const
             return {};
         }
         case Qt::ToolTipRole: {
+            if (column == ActiveColumn && !at(row).validMCMeta()) {
```
IMO it'd be better to have this on the name column (the same as the symlink warning) and have the messages stack

TheKodeToad commented:
Thank you! But I think a new PR and clean commit history would be better
